### PR TITLE
app/vminsert: dropSamplesOnOverload when storage node is ready

### PR DESCRIPTION
**Pull request by shichanglin5**

### CODE REVIEW

File: app/vminsert/netstorage/netstorage.go [hunk 1]
```diff
@@ -70,7 +70,7 @@ func (sn *storageNode) push(snb *storageNodesBucket, buf []byte, rows int) error
 		// Fast path - the buffer is successfully sent to sn.
 		return nil
 	}
-	if *dropSamplesOnOverload && !sn.isReadOnly.Load() {
+	if *dropSamplesOnOverload && !sn.isBroken.Load() && sn.isReadOnly.Load() {
```
Comment (f41gh7) [inline app/vminsert/netstorage/netstorage.go:73]: Could you please explain which problem solves this change? 

 Previously, samples could be dropped only if storage node is overloaded and it's not read only.

With this change, samples would be dropped on overload if storage is not broken and in read only state.

Comment (shichanglin5) [reply app/vminsert/netstorage/netstorage.go:73]: Sorry, there is a problem here, which leads to misleading.

**Comment by shichanglin5** [comment]
at `app/vminsert/netstorage/netstorage.go:783`
sn.isReady()  => !sn.isBroken.Load() && !sn.isReadOnly.Load()
```go
func (sn *storageNode) trySendBuf(buf []byte, rows int) bool {
	if !sn.isReady() {
		// Fast path without locking the sn.brLock.
		return false
	}

	sent := false
	sn.brLock.Lock()
	if sn.isReady() && len(sn.br.buf)+len(buf) <= maxBufSizePerStorageNode {
		sn.br.buf = append(sn.br.buf, buf...)
		sn.br.rows += rows
		sent = true
	}
	sn.brLock.Unlock()
	return sent
}
```
The prerequisite for overload should be sn.isReady

If storage node is offline, vm insert cannot establish connection with storage node.`readOnly` will always be`false `, at this time. if `dropSamplesOnOverload`  is `true`, all requests to this node will be dropped. 

however, in this case (storage is broken, readonly is false), it is expected to be reroute to other storage node.

**Comment by shichanglin5** [comment]
@f41gh7

**Review by f41gh7 — APPROVED** [review]
LGTM

 Please update changelog

**Review by hagen1778 — COMMENTED** [review]
@shichanglin5 @f41gh7 
The flag description explicitly states that it drops samples when storage node is slow or unavailable:
>  -dropSamplesOnOverload
     Whether to drop incoming samples if the destination vmstorage node is overloaded and/or unavailable. 

This change updates its behavior. What's the reasoning to not drop samples if node is down, but to drop samples if node is slow?

**Comment by shichanglin5** [comment]
> @shichanglin5 @f41gh7 The flag description explicitly states that it drops samples when storage node is slow or unavailable:
> 
> > -dropSamplesOnOverload
> > Whether to drop incoming samples if the destination vmstorage node is overloaded and/or unavailable.
> 
> This change updates its behavior. What's the reasoning to not drop samples if node is down, but to drop samples if node is slow?

the handling logic in the code for the overload and unavailable states appears to be separated. Please refer to the description in the screenshot below.
image
did I misunderstand it?
@hagen1778

**Comment by hagen1778** [comment]
> What's the reasoning to not drop samples if node is down, but to drop samples if node is slow?
> did I misunderstand it?

It is probably me who is explaining my concern poorly.

So, `-dropSamplesOnOverload` "drops incoming samples if the destination vmstorage node is **overloaded** and/or **unavailable**". There are 3 reasons why `sn.TrySendBuf` could fail:
1. remote storage is unavailable, and we get network error;
2. remote storage is degraded, and we get a timeout error;
3. remote storage tells us it is in RO mode.

In current logic, `if *dropSamplesOnOverload && !sn.isReadOnly.Load()` the rows will be dropped if `sn.TrySendBuf` failed and the reason of the failure is not p3 (RO mode).

In the new logic, `if *dropSamplesOnOverload && sn.isReady()` the rows will be dropped if `sn.TrySendBuf` failed and the reason of the failure is not p1 (network err) and not p3 (RO mode). This contradicts with flag description `if the destination vmstorage node is overloaded and/or unavailable`.

**Comment by shichanglin5** [comment]
> > What's the reasoning to not drop samples if node is down, but to drop samples if node is slow?
> > did I misunderstand it?
> 
> It is probably me who is explaining my concern poorly.
> 
> So, `-dropSamplesOnOverload` "drops incoming samples if the destination vmstorage node is **overloaded** and/or **unavailable**". There are 3 reasons why `sn.TrySendBuf` could fail:
> 
> 1. remote storage is **unavailable**, and we get network error;
> 2. remote storage is **overloaded**, and we get a timeout error;
> 3. remote storage tells us it is in **RO** mode.
> 
> In current logic, `if *dropSamplesOnOverload && !sn.isReadOnly.Load()` the rows will be dropped if `sn.TrySendBuf` failed and the reason of the failure is not p3 (**RO** mode).
> 
> In the new logic, `if *dropSamplesOnOverload && !sn.isBroken.Load() && !sn.isReadOnly.Load()` the rows will be dropped if `sn.TrySendBuf` failed and the reason of the failure is not p1 (**unavailable**, network err) and not p3 (**RO** mode). This contradicts with flag description `if the destination vmstorage node is overloaded and/or unavailable`.

Thanks! I understand what you mean, you explain the description of the `dropSamplesOnOverload` parameter in the code.

I want to explain that overload and broken are two different situations, and it is better to deal with them separately, so I think the change of the current branch  `dropSamplesOnOverload` is more reasonable.

Since you don't think it is necessary, I will close pr

**Pull request closed without merging**